botocore/stub.py

```diff
@@ -0,0 +1,164 @@
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import copy
+from botocore.validate import validate_parameters
+from botocore.exceptions import ParamValidationError, StubResponseError
+from botocore.vendored.requests.models import Response
+
+
+class QueueStubber(object):
+    """
+    This class will allow you to stub out requests so you don't have to hit
+    an endpoint to write tests. Responses are returned first in, first out.
+    If operations are called out of order, or are called with no remaining
+    queued responses, an error will be raised.
+    """
+    def __init__(self, client):
+        """
+        :param client: The client to add your stubs to.
+        """
+        self.client = client
+        self._queue = []
+
+    def activate(self):
+        """
+        Activates the stubber on the client
+        """
+        self.client.meta.events.register(
+            'before-call.*.*', self._get_response_handler)
+
+    def deactivate(self):
+        """
+        Deactivates the stubber on the client
+        """
+        self.client.meta.events.unregister(
+            'before-call.*.*', self._get_response_handler)
+
+    def add_response(self, method, service_response):
+        """
+        Adds a service response to the response queue. This will be validated
+        against the service model to ensure correctness. It should be noted,
+        however, that while missing attributes are often considered correct,
+        your code may not function properly if you leave them out. Therefore
+        you should always fill in every value you see in a typical response for
+        your particular request.
+
+        :param method: The name of the service method to stub.
+        :type method: str
+
+        :param service_response: A dict response stub. Provided parameters will
+            be validated against the service model.
+        :type service_response: dict
+        """
+        self._add_response(method, service_response)
+
+    def _add_response(self, method, service_response, http_response=None):
+        if not hasattr(self.client, method):
+            raise ValueError(
+                "Client %s does not have method: %s"
+                % (self.client.meta.service_model.service_name, method))
+
+        if http_response is None:
+            http_response = Response()
+            http_response.status_code = 200
+            http_response.reason = 'OK'
+
+        operation_name = self.client.meta.method_to_api_mapping.get(method)
+        self._validate_response(operation_name, service_response)
+
+        response = (operation_name, (http_response, service_response))
+        self._queue.append(response)
+
+    def add_client_error(self, method, service_error_code='',
+                         service_message='', http_status_code=400,
+                         http_reason=''):
```

Comment:
What's the reason for including the `http_reason`?  This is tied to the http status code and isn't something the user can generally override is it?

If we find this useful to have we can automatically lookup the reason for a given status code (I believe it's somewhere either in requests/urllib3/httplib), but it doesn't seem like the user should be able to provide this.